### DISCUSSION TITLE
How to use the new Update and Insert Queries from v0.29

### DISCUSSION THREAD

**Discussion by pcriv** [stephenafamo/bob#308] [Q&A] [answered]
I'm trying to find out how could I convert this queries that were working on v0.28 to the new approach of version v0.29

Update:
```go
_, err = dbgen.Blocks.UpdateQ(ctx, bob.New(m.db), &dbgen.BlockSetter{CRC32: omitnull.From(crc32Bytes)},dbgen.UpdateWhere.Blocks.Key.EQ(block.Key)).Exec()
```

Insert:
```go
_, err = dbgen.Blocks.Insert(ctx, bob.New(m.db), &dbgen.BlockSetter{
  Key:      omit.From(u.Key),
  State:    omit.From(BlockExpected.Int32()),
  Bytesize: omitnull.From(u.ByteSize),
  BucketID: omit.From(DefaultBucketID),
})
```

**Comment by pcriv** [comment, marked as answer]
Okay I think I found the way:

Update:

```go
_, err = dbgen.Blocks.Update(
    dbgen.BlockSetter{CRC32: omitnull.From(crc32Bytes)}.UpdateMod(),
    dbgen.UpdateWhere.Blocks.Key.EQ(block.Key),
).Exec(ctx, bob.New(m.db))
```
Insert:

```go
_, err = dbgen.Blocks.Insert(&dbgen.BlockSetter{
    Key:      omit.From(u.Key),
    State:    omit.From(BlockExpected.Int32()),
    Bytesize: omitnull.From(u.ByteSize),
    BucketID: omit.From(DefaultBucketID),
}).One(ctx, bob.New(m.db))
```

**Comment by Quip1995** [comment]
This was a godsend for someone just getting started with this project. Could some of that context not be derived from the individual model? When I tried doing this via the individual model on the docs ([here](https://bob.stephenafamo.com/docs/code-generation/usage#update)), I would receive an error: `pq: missing FROM-clause entry for table`

**Reply by stephenafamo** [reply]
Can you share more details about this? What does the model look like? What does your code look like?


Also use `bob.Debug` to print out the queries run so that I can take a look.

**Reply by Quip1995** [reply]
So the [Documentation](https://bob.stephenafamo.com/docs/code-generation/usage) appears so suggest that the following should work: 

```go
//get the row from the database using generated models and pk
customer, err := generatedModels.FindCustomer(ctx, bobExec, 1)
if err != nil {
	return nil, err
}

//update the row
err = customer.Update(ctx, bobExec, &generatedModels.CustomerSetter{})
if err != nil {
	return nil, err
}
```

But that leads to this error: 
> pq: missing FROM-clause entry for table "customer"

Using bob.debug I get (I omitted the RETURNING value): 

> UPDATE "customer" AS "customer" SET
"name" = $1
WHERE ("customer"."id" = $2)
RETURNING  ...

So I can see that it's specifying a where clause, it just doesn't seem to be extracting it via the method. I need to update my code to the following 

```go

_, err = generatedModels.Customer.Update(
	generatedModels.CustomerSetter{
		//Setters
	}.UpdateMod(),
	generatedModels.UpdateWhere.Customer.ID.EQ(1),
).Exec(context.TODO(), bobExec)
if err != nil {
	return fmt.Errorf("failed to update location: %w", err)
}

```

And then the update works just fine. Perhaps I'm doing something wrong?

Obviously the downside in doing this is that EXEC doesn't return the full model reflecting the updates, so I need to retrieve again if I want to ensure that my changes are there.
